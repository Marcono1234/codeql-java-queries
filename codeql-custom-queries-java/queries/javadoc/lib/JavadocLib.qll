
import java

bindingset[s, toCount, endIndexInclusive]
private int countToIndex(string s, string toCount, int endIndexInclusive) {
    result = count(int index |
        index = s.indexOf(toCount)
        and index <= endIndexInclusive
    |
        index
    )
}

bindingset[s]
private int getInlineTagEndIndexExclusive(string s) {
    // Count opening and closing curly brackets, see also https://docs.oracle.com/en/java/javase/16/docs/specs/javadoc/doc-comment-spec.html#inline-tags
    result = min(int balancedIndex, int openingCount, int closingCount |
        balancedIndex = [0 .. s.length() - 1]
        and openingCount = countToIndex(s, "{", balancedIndex)
        and closingCount = countToIndex(s, "}", balancedIndex)
        and openingCount = closingCount
    |
        // + 1 to make index exclusive
        balancedIndex + 1
    )
}

private string getInlineTagPattern(boolean lazy) {
    exists(string quantifier |
        lazy = true and quantifier = "?"
        or lazy = false and quantifier = ""
    |    
        // Even though (some?) inline tags can span multiple lines, don't check multiline for now
        /*
        * Exact format is not specified, and parsing also seems to differ between tag types, see
        * https://github.com/openjdk/jdk/blob/4eacdb38a83b545603928392eccb116c744ef3b3/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java
        * The following pattern matches how some standard JDK tags are parsed
        * 
        * Group 1: tag name
        * Group 2: tag content
        */
        result = "(?s)\\{@([a-zA-Z0-9\\-_.:]+)\\p{javaWhitespace}+(.*" + quantifier + ")(?<!\\p{javaWhitespace})\\p{javaWhitespace}*\\}"
    )
}

bindingset[javadoc]
private string getInlineTagContent_(string javadoc, string tagName, int index, int length) {
    exists(string potentialInlineTag, string actualInlineTag |
        // For finding tag use lazy pattern, to not match multiple inline tags at once
        exists(javadoc.regexpFind(getInlineTagPattern(true), _, index))
        and potentialInlineTag = javadoc.suffix(index)
        // Get the actual end, this is at the same time the length of the tag (including closing '}')
        and length = getInlineTagEndIndexExclusive(potentialInlineTag)
        and actualInlineTag = potentialInlineTag.prefix(length)
        // For capturing use greedy pattern
        and exists(string capturePattern | capturePattern = getInlineTagPattern(false) |
            tagName = actualInlineTag.regexpCapture(capturePattern, 1)
            and result = actualInlineTag.regexpCapture(capturePattern, 2)
        )
    )
}

/**
 * Gets the content of an inline Javadoc tag. The complete tag, including opening
 * and closing curly brackets, starts at `index` and has length `length`.
 * Has no result if the string does not contain any inline tags, the inline tags have
 * no content or are malformed.
 */
bindingset[javadoc]
string getInlineTagContent(string javadoc, string tagName, int index, int length) {
    result = getInlineTagContent_(javadoc, tagName, index, length)
    /*
     * And because finding is done lazily, make sure that there is no enclosing inline tag
     * because nested inline tags are treated literally (only for some kinds?)
     * 
     * This requires having the actual implementation in separate predicate `getInlineTagContent_`,
     * otherwise it would lead to non-monotonic recursion (which is not permitted)
     * 
     * No need to check this recursively because brackets have to be balanced, so even
     * if the enclosing tag has itself an enclosing tag, then that would cover all tags
     */
    and not exists(int enclosingIndex, int enclosingLength |
        exists(getInlineTagContent_(javadoc, _, enclosingIndex, enclosingLength))
        and index = [enclosingIndex + 1 .. enclosingIndex + enclosingLength - 1]
    )
}

bindingset[maxIndexExclusive]
private string getCompleteJavadocTextUpTo(JavadocParent parent, int maxIndexExclusive) {
    // JavadocText only represents a single line, see https://github.com/github/codeql/issues/3696
    // Does not work correctly for Javadoc block tags, see https://github.com/github/codeql/issues/3825
    result = concat(int index, string line |
        line = parent.getChild(index).(JavadocText).getText()
        and index < maxIndexExclusive
    |
        line, "\n" order by index asc
    )
}

/**
 * Gets the complete Javadoc text of the parent, that is the text of all
 * lines concatenated with `\n`.
 */
string getCompleteJavadocText(JavadocParent parent) {
    result = getCompleteJavadocTextUpTo(parent, 2147483647)
}

/**
 * Gets the start index (starting at 0) of the JavadocText in the complete
 * text of the parent (as created by `getCompleteJavadocText`).
 */
int getStartIndexInCompleteJavadocText(JavadocParent parent, JavadocText javadocText) {
    exists(int index, int offset | parent.getChild(index) = javadocText |
        (
            index = 0 and offset = 0
            // + 1 for trailing line terminator, see `getCompleteJavadocTextUpTo` implementation
            or index > 0 and offset = 1
        )
        and result = getCompleteJavadocTextUpTo(parent, index).length() + offset
    )
}
