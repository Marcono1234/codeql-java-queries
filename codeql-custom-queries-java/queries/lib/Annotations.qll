import java

// Hacky workaround to prevent Class from choosing inherited annotations
private class DeclaredAnnotationsRetriever extends Class, RefType {
    override
    Annotation getAnAnnotation() {
        result = RefType.super.getAnAnnotation()
    }
}

/**
 * Gets a declared annotation of the annotated element. Unlike `Annotatable.getAnAnnotation()`
 * this predicate does not have inherited annotations as result.
 */
Annotation getADeclaredAnnotation(Annotatable annotated) {
    if (annotated instanceof Class) then result = annotated.(DeclaredAnnotationsRetriever).getAnAnnotation()
    else result = annotated.getAnAnnotation()
}

bindingset[targetType]
predicate isApplicableToTargetType(AnnotationType annotationType, string targetType) {
    if (annotationType.getAnAnnotation() instanceof TargetAnnotation) then (
        targetType = annotationType.getAnAnnotation().(TargetAnnotation).getATargetElementType()
    ) else (
        // No Target annotation means "applicable to all contexts" since JDK 14, see https://bugs.openjdk.java.net/browse/JDK-8231435
        // The compiler does not completely implement that, but pretend it did
        any()
    )
}

/**
 * Gets a string which describes the target types of the annotation type in a
 * human-readable format.
 */
string describeTargetTypes(AnnotationType annotationType) {
    if (annotationType.getAnAnnotation() instanceof TargetAnnotation) then (
        result = concat(annotationType.getAnAnnotation().(TargetAnnotation).getATargetElementType(), ", ")
    ) else (
        // No Target annotation means "applicable to all contexts" since JDK 14, see https://bugs.openjdk.java.net/browse/JDK-8231435
        // The compiler does not completely implement that, but pretend it did
        result = "<all>"
    )
}

class RepeatableAnnotation extends Annotation {
    RepeatableAnnotation() {
        getType().hasQualifiedName("java.lang.annotation", "Repeatable")
    }

    /**
     * Gets the annotation type which acts as _containing type_, grouping multiple
     * repeatable annotations together.
     */
    AnnotationType getContainingType() {
        // Get annotation type A of Class<A> specified by `value`
        result = getValue("value").(TypeLiteral).getTypeName().getType()
    }
}

class DocumentedAnnotation extends Annotation {
    DocumentedAnnotation() {
        getType().hasQualifiedName("java.lang.annotation", "Documented")
    }
}

class InheritedAnnotation extends Annotation {
    InheritedAnnotation() {
        getType().hasQualifiedName("java.lang.annotation", "Inherited")
    }
}
