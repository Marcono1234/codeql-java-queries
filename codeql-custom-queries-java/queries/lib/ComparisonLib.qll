import java

/**
 * A comparison method which can be implemented, this includes methods such as `Comparable.compareTo`
 * and `Comparator.compare`.
 */
class ImplementableComparisonMethod extends Method {
    ImplementableComparisonMethod() {
        getDeclaringType().hasQualifiedName("java.lang", "Comparable") and hasName("compareTo")
        or getDeclaringType().hasQualifiedName("java.util", "Comparator") and hasName("compare")
        or getDeclaringType().hasQualifiedName("java.text", "Collator") and hasName("compare")
    }
}

/**
 * A method which compares two values and returns an `int` representing the comparison result.
 */
class ComparisonMethod extends Method {
    ComparisonMethod() {
        getSourceDeclaration().getASourceOverriddenMethod+() instanceof ImplementableComparisonMethod
        // Or a static comparison method
        or exists(RefType declaringType |
            declaringType = getDeclaringType()
        |
            declaringType instanceof BoxedType and hasName(["compare", "compareUnsigned"])
            or declaringType.hasQualifiedName("java.lang", "CharSequence") and hasName("compare")
            or declaringType.hasQualifiedName("java.util", "Arrays") and hasName(["compare", "compareUnsigned"])
            or declaringType.hasQualifiedName("java.util", "Objects") and hasName("compare")
        )
    }

    /**
     * Holds if this comparison method specifies exact return values (e.g. -1, 0 and 1) instead
     * of only defining the sign of the result.
     */
    predicate definesSpecificReturnValues() {
        getDeclaringType().getASourceSupertype*().hasQualifiedName("java.math", ["BigDecimal", "BigInteger"])
    }
}
