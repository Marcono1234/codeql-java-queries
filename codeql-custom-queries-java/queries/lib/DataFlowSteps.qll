import java
import semmle.code.java.dataflow.DataFlow

private predicate isOwnFieldAccess(FieldAccess fieldAccess) {
    fieldAccess.getField().isStatic()
    or fieldAccess.isOwnFieldAccess()
}

/**
 * Holds if the step from `node1` to `node2` is the assignment to a field followed
 * by a read of the field.
 */
predicate isOwnFieldStep(DataFlow::Node node1, DataFlow::Node node2) {
    exists(FieldWrite fieldWrite, FieldRead fieldRead |
        fieldWrite.getField() = fieldRead.getField()
        and isOwnFieldAccess(fieldWrite)
        and isOwnFieldAccess(fieldRead)
        and fieldWrite.getRhs() = node1.asExpr()
        and fieldRead = node2.asExpr()
    )
}
