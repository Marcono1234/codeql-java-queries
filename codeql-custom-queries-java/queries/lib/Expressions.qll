import java

/**
 * A statement expression, as specified by [JLS 16 ยง14.8](https://docs.oracle.com/javase/specs/jls/se16/html/jls-14.html#jls-14.8).
 * The result value of a statement expression, if any, is discarded.
 *
 * Not to be confused with `ExprStmt`; while the child of an `ExprStmt` is always
 * a `StmtExpr` the opposite is not true, a `StmtExpr` occurs for example also
 * as 'init' of a `for` statement.
 */
class StmtExpr extends Expr {
    StmtExpr() {
        this = any(ExprStmt s).getExpr()
        or exists(ForStmt forStmt |
            this = forStmt.getAnInit()
            or this = forStmt.getAnUpdate()
        )
        // Only applies to SwitchStmt, but not to SwitchExpr
        or this = any(SwitchStmt s).getACase().getRuleExpression()
        // TODO: Workarounds for https://github.com/github/codeql/issues/3605
        or exists(LambdaExpr lambda |
            this = lambda.getExprBody()
            and lambda.asMethod().getReturnType() instanceof VoidType
        )
        or exists(MemberRefExpr memberRef, Method implicitMethod, Method overridden |
            implicitMethod = memberRef.asMethod()
        |
            this.getParent().(ReturnStmt).getEnclosingCallable() = implicitMethod
            // asMethod() has bogus method with wrong return type as result, e.g. `run(): String` (overriding `Runnable.run(): void`)
            // Therefore need to check the overridden method
            and implicitMethod.getSourceDeclaration().overridesOrInstantiates*(overridden)
            and overridden.getReturnType() instanceof VoidType
        )
    }
}

/**
 * A text block (Java 15 feature).
 */
class TextBlock extends StringLiteral {
    // Detection does not cover all cases, but most cases
    // Ideally this would be part of the standard CodeQL classes, see https://github.com/github/codeql/issues/6619
    TextBlock() {
        // Literal contains line terminator; this is always the case behind the starting """
        // and it is not possible for regular string literals, there line terminator
        // would have to be escaped
        // Would not work if line terminator uses Unicode escape
        getLiteral().matches(["%\n%", "%\r%"])
    }

    /**
     * Gets the line of the literal at index `lineIndex` relative to the first line
     * (starting at 0) as it appeared in source. The opening line containing `"""`
     * followed by a line terminator is not part of the result set.
     * 
     * Trailing backslashes in lines (to continue in the next line) are not treated
     * specially because they are processed during regular escape sequence replacement,
     * after incidental whitespaces have been removed.
     */
    string getLiteralLine(int lineIndex) {
        // Note: Won't work correctly when closing """ uses Unicode escapes or when
        // line terminators use Unicode escapes
        exists(string literal, int index, string rawLine | literal = getLiteral() |
            // lineIndex + 1 because the first line contains no content, only the opening """
            rawLine = literal.regexpFind("(?m)^.*$", lineIndex + 1, index)
            and if (index + rawLine.length() = literal.length()) then (
                // Remove trailing """
                result = rawLine.prefix(rawLine.length() - 3)
            ) else (
                result = rawLine
            )
        )
        // Ignore line containing opening """
        and lineIndex != -1
    }
}