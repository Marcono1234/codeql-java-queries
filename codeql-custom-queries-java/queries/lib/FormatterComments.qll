import java
import semmle.code.xml.XML

// XML CodeQL classes don't extend Top, must use algebraic datatype as workaround
// See also https://github.com/github/codeql/issues/5646

private newtype TFormatterComment = TJavaComment(JavadocText t) or TXmlComment(XMLComment c)

/**
 * A comment which enables or disables the IDE formatter or a formatter plugin for a
 * range of code.
 */
abstract class FormatterComment extends TFormatterComment {
    /**
     * Gets the prefix of the comment, differentiating between comments for different
     * formatters
     */
    abstract string getPrefix();

    /**
     * Gets whether this comment is turning the formatter on (`true`) or off (`false`).
     */
    abstract boolean getIsFormatterOn();

    abstract Location getLocation();

    int getLineNumber() {
        result = getLocation().getStartLine()
    }

    /**
     * Gets a comment with the same prefix in the same type as this one.
     */
    FormatterComment getACommentInSameFile() {
        result.getLocation().getFile() = getLocation().getFile()
        and result.getPrefix() = getPrefix()
        and result != this
    }

    abstract string toString();
}

bindingset[comment, commentPrefix]
private boolean getIsCommentFormatterOn(string comment, string commentPrefix) {
    result = true and comment.matches("%" + commentPrefix + ":on%")
    or result = false and comment.matches("%" + commentPrefix + ":off%")
}

/**
 * Eclipse and IntelliJ IDEA Java source formatter comment
 */
private class EclipseIntelliJFormatterComment extends FormatterComment, TJavaComment {
    JavadocText comment;
    boolean isOn;

    EclipseIntelliJFormatterComment() {
        this = TJavaComment(comment)
        // Make sure this is not an actual Javadoc comment, see https://github.com/github/codeql/issues/3695
        and not exists(comment.getJavadoc().getCommentedElement())
        and isOn = getIsCommentFormatterOn(comment.getText(), "@formatter")
    }

    override
    string getPrefix() {
        result = "@formatter"
    }

    override
    boolean getIsFormatterOn() {
        result = isOn
    }

    override
    Location getLocation() {
        result = comment.getLocation()
    }

    override
    string toString() {
        result = comment.toString()
    }
}

/**
 * Spotless Java source formatter comment, see
 * https://github.com/diffplug/spotless/tree/main/plugin-gradle#spotlessoff-and-spotlesson
 * 
 * Spotless does not detect mismatching comments yet, see https://github.com/diffplug/spotless/issues/1165
 */
private class SpotlessJavaFormatterComment extends FormatterComment, TJavaComment {
    JavadocText comment;
    boolean isOn;

    SpotlessJavaFormatterComment() {
        this = TJavaComment(comment)
        // Make sure this is not an actual Javadoc comment, see https://github.com/github/codeql/issues/3695
        and not exists(comment.getJavadoc().getCommentedElement())
        and isOn = getIsCommentFormatterOn(comment.getText(), "spotless")
    }

    override
    string getPrefix() {
        result = "spotless"
    }

    override
    boolean getIsFormatterOn() {
        result = isOn
    }

    override
    Location getLocation() {
        result = comment.getLocation()
    }

    override
    string toString() {
        result = comment.toString()
    }
}

/**
 * Spotless XML formatter comment, see
 * https://github.com/diffplug/spotless/tree/main/plugin-gradle#spotlessoff-and-spotlesson
 * 
 * Spotless does not detect mismatching comments yet, see https://github.com/diffplug/spotless/issues/1165
 */
private class SpotlessXmlFormatterComment extends FormatterComment, TXmlComment {
    XMLComment comment;
    boolean isOn;

    SpotlessXmlFormatterComment() {
        this = TXmlComment(comment)
        and isOn = getIsCommentFormatterOn(comment.getText(), "spotless")
    }

    override
    string getPrefix() {
        result = "spotless"
    }

    override
    boolean getIsFormatterOn() {
        result = isOn
    }

    override
    Location getLocation() {
        result = comment.getLocation()
    }

    override
    string toString() {
        result = comment.toString()
    }
}
