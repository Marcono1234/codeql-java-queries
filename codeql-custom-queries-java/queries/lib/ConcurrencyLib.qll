import java

import lib.VarAccess

// TODO: Maybe also consider java.util.concurrent.locks.StampedLock

/**
 * Some form of synchronization which guards zero or more statements.
 */
abstract class Synchronization extends Top {
    /**
     * Holds if this synchronization contains the given expression.
     */
    abstract predicate includesStmt(Stmt stmt);
    
    /**
     * Holds if this synchronization contains the given expression.
     */
    abstract predicate includes(Expr expr);
    
    /**
     * Holds if this synchronization uses the same lock as the other one.
     */
    abstract predicate usesSameLockAs(Synchronization other);
    
    /**
     * Gets a human-readable string describing the kind of synchronization.
     */
    abstract string describe();
}

pragma[inline] // inline this, otherwise this is extremely inefficient
private predicate haveCommonAncestor(RefType a, RefType b) {
    a.getASourceSupertype*() = b.getASourceSupertype*()
}

private class SynchronizedStmt_ extends Synchronization, SynchronizedStmt {
    override predicate includesStmt(Stmt stmt) {
        stmt.getEnclosingStmt().getEnclosingStmt*() = this
    }

    override predicate includes(Expr expr) {
        includesStmt(expr.getEnclosingStmt())
    }
    
    override predicate usesSameLockAs(Synchronization other) {
        haveCommonAncestor(getExpr().(ThisAccess).getType(), other.(SynchronizedStmt_).getExpr().(ThisAccess).getType())
        or haveCommonAncestor(getExpr().(ThisAccess).getType(), other.(SynchronizedMethod).getDeclaringType())
        or accessSameVarOfSameOwner(getExpr(), other.(SynchronizedStmt_).getExpr())
        or getExpr().(TypeLiteral).getType() = other.(SynchronizedStmt).getExpr().(TypeLiteral).getType()
        or getExpr().(TypeLiteral).getType() = other.(SynchronizedStaticMethod).getDeclaringType()
    }
    
    override string describe() {
        result = "synchronized statement"
    }
}

private class SynchronizedMethod extends Synchronization, Method {
    SynchronizedMethod() {
        not isStatic()
        and isSynchronized()
    }
    
    override predicate includesStmt(Stmt stmt) {
        stmt.getEnclosingCallable() = this
    }

    override predicate includes(Expr expr) {
        includesStmt(expr.getEnclosingStmt())
    }
    
    override predicate usesSameLockAs(Synchronization other) {
        haveCommonAncestor(getDeclaringType(), other.(SynchronizedMethod).getDeclaringType())
        // Covers SynchronizedStmt_ checks
        or other.usesSameLockAs(this)
    }
    
    override string describe() {
        result = "synchronized method"
    }
}

private class SynchronizedStaticMethod extends Synchronization, Method {
    SynchronizedStaticMethod() {
        isStatic()
        and isSynchronized()
    }
    
    override predicate includesStmt(Stmt stmt) {
        stmt.getEnclosingCallable() = this
    }

    override predicate includes(Expr expr) {
        includesStmt(expr.getEnclosingStmt())
    }
    
    override predicate usesSameLockAs(Synchronization other) {
        getDeclaringType() = other.(SynchronizedStaticMethod).getDeclaringType()
        // Covers SynchronizedStmt_ checks
        or other.usesSameLockAs(this)
    }
    
    override string describe() {
        result = "synchronized static method"
    }
}

private class TypeLock extends Interface {
    TypeLock() {
        hasQualifiedName("java.util.concurrent.locks", "Lock")
    }
}

private class LockMethod extends Method {
    LockMethod() {
        getAnOverride*().getDeclaringType() instanceof TypeLock
        and hasName(["lock", "lockInterruptibly", "tryLock"])
    }
}

private class UnlockMethod extends Method {
    UnlockMethod() {
        getAnOverride*().getDeclaringType() instanceof TypeLock
        and hasName("unlock")
    }
}

private class TypeReadWriteLock extends Interface {
    TypeReadWriteLock() {
        hasQualifiedName("java.util.concurrent.locks", "ReadWriteLock")
    }
}

private class ReadLockMethod extends Method {
    ReadLockMethod() {
        getAnOverride*().getDeclaringType() instanceof TypeReadWriteLock
        and hasName("readLock")
    }
}

private class WriteLockMethod extends Method {
    WriteLockMethod() {
        getAnOverride*().getDeclaringType() instanceof TypeReadWriteLock
        and hasName("writeLock")
    }
}

private class LockMethodCalls extends Synchronization, MethodAccess {
    LockMethodCalls() {
        getMethod() instanceof LockMethod
    }

    private predicate isSameLock(Expr a, Expr b) {
        accessSameVarOfSameOwner(a, b)
        // Or usage of ReadWriteLock locks
        or exists(Method readWriteLockGetter |
            readWriteLockGetter instanceof ReadLockMethod
            or readWriteLockGetter instanceof WriteLockMethod
        |
            a.(MethodAccess).getMethod() = readWriteLockGetter
            and b.(MethodAccess).getMethod() = readWriteLockGetter
            and accessSameVarOfSameOwner(a.(MethodAccess).getQualifier(), b.(MethodAccess).getQualifier())
        )
    }
    
    override predicate includesStmt(Stmt stmt) {
        exists (MethodAccess unlockCall |
            unlockCall.getMethod() instanceof UnlockMethod
            and isSameLock(getQualifier(), unlockCall.getQualifier())
        |
            this.getControlFlowNode().getASuccessor+() = stmt
            and stmt.getControlFlowNode().getASuccessor+() = unlockCall
        )
    }

    override predicate includes(Expr expr) {
        exists (MethodAccess unlockCall |
            unlockCall.getMethod() instanceof UnlockMethod
            and isSameLock(getQualifier(), unlockCall.getQualifier())
        |
            this.getControlFlowNode().getASuccessor+() = expr
            and expr.getControlFlowNode().getASuccessor+() = unlockCall
        )
    }
    
    override predicate usesSameLockAs(Synchronization other) {
        isSameLock(getQualifier(), other.(LockMethodCalls).getQualifier())
    }
    
    override string describe() {
        if hasQualifier() then (
            result = "lock on " + getQualifier()
        ) else (
            result = "own lock methods"
        )
    }
}

/**
 * Some kind of statement which marks the start of a synchronized section, or the complete
 * synchronized section (in case of `synchronized` statements).
 */
class SynchronizationStatement extends Synchronization, Stmt {
    Synchronization delegate;

    SynchronizationStatement() {
        delegate.(SynchronizedStmt_) = this
        or
        delegate.(LockMethodCalls).getEnclosingStmt() = this
    }

    override predicate includesStmt(Stmt stmt) {
        delegate.includesStmt(stmt)
    }

    override predicate includes(Expr expr) {
        delegate.includes(expr)
    }
    
    override predicate usesSameLockAs(Synchronization other) {
        delegate.usesSameLockAs(other)
    }
    
    override string describe() {
        result = delegate.describe()
    }
}

/**
 * Holds if the expression is synchronized by the specified synchronization, or a
 * synchronization which uses the same lock.
 */
predicate isExprSynchronizedBy(Expr expr, Synchronization synchronization) {
    synchronization.includes(expr)
    or exists (Synchronization other | other.usesSameLockAs(synchronization) |
        other.includes(expr)
    )
}

predicate isExprSynchronized(Expr e) {
    isExprSynchronizedBy(e, _)
}

/**
 * Holds if safe publication is required for instances of the class to ensure that other
 * threads see an up-to-date version of the instances.
 * 
 * See also [JLS: `final` Field Semantics](https://docs.oracle.com/javase/specs/jls/se17/html/jls-17.html#jls-17.5)
 */
predicate requiresSafePublication(Class c) {
    exists (Field f | f = c.getAField() |
        not f.isStatic()
        and not (f.isFinal() or f.isVolatile())
    )
    or exists (RefType superType | superType = c.getASourceSupertype() |
        requiresSafePublication(superType)
    )
}
