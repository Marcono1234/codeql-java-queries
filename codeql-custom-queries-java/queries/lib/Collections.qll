import java
import semmle.code.java.dataflow.DataFlow

import lib.Expressions

/**
 * The interface `java.util.Collection`.
 */
class TypeCollection extends Interface {
    TypeCollection() {
        hasQualifiedName("java.util", "Collection")
    }
}

/**
 * The interface `java.util.Set`.
 */
class TypeSet extends Interface {
    TypeSet() {
        hasQualifiedName("java.util", "Set")
    }
}

/**
 * `Set` subtype which preserves the insertion order.
 */
class InsertionOrderSetType extends RefType {
    InsertionOrderSetType() {
        hasQualifiedName("java.util", "LinkedHashSet")
        or hasQualifiedName("java.util.concurrent", "CopyOnWriteArraySet")
    }
}

/**
 * `Set` subtype which orders its elements in some consistent way.
 */
class OrderedSetType extends RefType {
    OrderedSetType() {
        this instanceof InsertionOrderSetType
        or hasQualifiedName("java.util", [
            "EnumSet",
            "SortedSet",
        ])
    }
}

/**
 * Holds if the expression represents access to a collection over which a
 * (possibly indirect) iteration is performed which preserves the order
 * of the elements.
 * 
 * If `includeSynchronized` is `true` this includes methods which might
 * synchronize on the collection.
 */
bindingset[includeSynchronized]
predicate isOrderPreservingCollectionIteration(Expr e, boolean includeSynchronized) {
    // Iterates (implicitly) over the elements
    exists(CallableReferencingExpr callableReferencingExpr, string methodName |
        callableReferencingExpr.getQualifier() = e
        and methodName = callableReferencingExpr.getReferencedCallable().(Method).getName()
        and (
            methodName = [
                "iterator",
                "parallelStream",
                "spliterator",
                "stream",
            ]
            or
            includeSynchronized = true and methodName = [
                "forEach",
                "toArray",
            ]
        )
    )
    // Or creates a new Collection which preserves the order of the elements
    or exists(ClassInstanceExpr newCollectionExpr, RefType constructedType |
        constructedType = newCollectionExpr.getConstructedType().getSourceDeclaration()
        and (
            constructedType.getASourceSupertype*() instanceof InsertionOrderSetType
            // Any List or Queue implementation preserves order
            or constructedType.getASourceSupertype*().hasQualifiedName("java.util", ["List", "Queue"])
        )
        and newCollectionExpr.getAnArgument() = e
    )
    // Or used in enhanced `for` statement which implicitly calls `iterator()`
    or exists(EnhancedForStmt forStmt | forStmt.getExpr() = e)
}

/**
 * Holds if the expression represents access to a collection over which a
 * (possibly indirect) iteration is performed.
 */
bindingset[includeSynchronized]
predicate isCollectionIteration(Expr e, boolean includeSynchronized) {
    isOrderPreservingCollectionIteration(e, includeSynchronized)
    // Or creates a new Collection of any type
    or exists(ClassInstanceExpr newCollectionExpr |
        newCollectionExpr.getConstructedType().getSourceDeclaration().getASourceSupertype*() instanceof TypeCollection
        and newCollectionExpr.getAnArgument() = e
    )
}

/**
 * The interface `java.util.Map`.
 */
class TypeMap extends Interface {
    TypeMap() {
        hasQualifiedName("java.util", "Map")
    }
}

/**
 * `Map` subtype which preserves the insertion order.
 */
class InsertionOrderMapType extends RefType {
    InsertionOrderMapType() {
        hasQualifiedName("java.util", "LinkedHashMap")
    }
}

/**
 * `Map` subtype which orders its entries in some consistent way.
 */
class OrderedMapType extends RefType {
    OrderedMapType() {
        this instanceof InsertionOrderMapType
        or hasQualifiedName("java.util", [
            "EnumMap",
            "SortedMap",
        ])
    }
}

/**
 * Holds if the expression represents access to a map over which a
 * (possibly indirect) iteration is performed which preserves the order
 * of the entries.
 * 
 * If `includeSynchronized` is `true` this includes methods which might
 * synchronize on the map.
 */
bindingset[includeSynchronized]
predicate isOrderPreservingMapIteration(Expr e, boolean includeSynchronized) {
    // Iterates (implicitly) over the entries
    exists(CallableReferencingExpr callableReferencingExpr, string methodName |
        callableReferencingExpr.getQualifier() = e
        and methodName = callableReferencingExpr.getReferencedCallable().(Method).getName()
    |
        includeSynchronized = true and methodName = [
            "forEach",
        ]
    )
    // Or creates a new map which preserves the entry insertion order
    or exists(ClassInstanceExpr newMapExpr |
        newMapExpr.getConstructedType().getSourceDeclaration().getASourceSupertype*() instanceof InsertionOrderMapType
        and newMapExpr.getAnArgument() = e
    )
}

/**
 * Holds if the expression represents access to a map over which a
 * (possibly indirect) iteration is performed which preserves the order
 * of the entries.
 */
bindingset[includeSynchronized]
predicate isMapIteration(Expr e, boolean includeSynchronized) {
    isOrderPreservingMapIteration(e, includeSynchronized)
    // Or creates a new map of any type
    or exists(ClassInstanceExpr newMapExpr |
        newMapExpr.getConstructedType().getSourceDeclaration().getASourceSupertype*() instanceof TypeMap
        and newMapExpr.getAnArgument() = e
    )
}

/**
 * Holds if the step from `node1` to `node2` is the access of the key, value or
 * entry collection of a map.
 */
predicate isMapCollectionStep(DataFlow::Node node1, DataFlow::Node node2) {
    exists(MethodAccess call |
        call.getMethod().hasName([
            "entrySet",
            "keySet",
            "values",
        ])
        and node1.asExpr() = call.getQualifier()
        and node2.asExpr() = call
    )
}

/**
 * Holds if the step from `node1` to `node2` is the creation of "sub collections", e.g. a
 * call to `List.subList` or similar.
 */
predicate isSubCollectionStep(DataFlow::Node node1, DataFlow::Node node2) {
    exists(MethodAccess call, Method baseMethod, RefType resultSupertype |
        // Check the return type of the original method to reduce false positives for `List<List<...>>` or similar
        // where element access methods such as `get()` return a collection as well
        baseMethod = call.getMethod().getSourceDeclaration().getASourceOverriddenMethod*()
        and not exists(baseMethod.getASourceOverriddenMethod())
        and resultSupertype = baseMethod.getReturnType().(RefType).getSourceDeclaration().getASourceSupertype*()
        and (
            resultSupertype instanceof TypeCollection
            or resultSupertype instanceof TypeMap
        )
        and node1.asExpr() = call.getQualifier()
        and node2.asExpr() = call
    )
}
