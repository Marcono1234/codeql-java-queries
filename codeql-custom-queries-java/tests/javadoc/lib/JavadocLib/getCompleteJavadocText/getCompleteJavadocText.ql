import java
import javadoc.lib.JavadocLib

from JavadocParent parent
where exists(Javadoc javadoc | javadoc = parent or javadoc.getAChild+() = parent |
    exists(javadoc.getCommentedElement())
)
select parent, getCompleteJavadocText(parent)
