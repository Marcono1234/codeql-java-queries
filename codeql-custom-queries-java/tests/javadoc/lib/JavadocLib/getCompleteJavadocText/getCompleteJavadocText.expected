| Test.java:2:5:5:7 | /** First line ... */ | First line\nsecond line |
| Test.java:8:5:8:10 | /**  */ |  |
| Test.java:12:5:16:7 | /** First ... */ | First\nDeprecated2 |
| Test.java:14:8:14:19 | @deprecated | Deprecated1 |
